## VCF indel notation converter from (A -> "-") to (GA -> G) including leading base (proper VCF notation).

#### For info please contact me: or.yaacov@mail.huji.ac.il
##### Optional: a more advanced script in this repo, dbSNPmatchindels.r, matches the indels with dbSNP.

##### Dependencies:
R (3+), 
Bioconductor (packages: BSgenomem, BSgenome.Hsapiens.UCSC.hg19, Biostrings)


Takes a 5 col tsv file (chr, pos, name, ref, alt):
>chr1	20996757	NULL	T	- <br>
>chr1	20996257	NULL	TT	- <br>
>chr1	20996457	NULL	-	TT <br>
>chr1	20996457	NULL	-	T <br>
>chr1	20996457	NULL	A	G<br>

Converts to:
>chr1	20996756	NULL	AT	A<br>
>chr1	20996256	NULL	GTT	G<br>
>chr1	20996456	NULL	A	ATT<br>
>chr1	20996456	NULL	A	AT<br>
>chr1	20996457	NULL	A	G<br>

#### Limitation: The script assumes "true" alternative and reference alleles (proper VCF), and not major/minor as in some files generated by PLINK, since in the case of major/minor it is usually impossible to distinguish a deletion from an insertion.
